### PR TITLE
Improved thread safety and convenience

### PR DESCRIPTION
**Pull request by oblivia-simplex**
This PR contains two sets of fairly minor changes:

1. I made a few changes to the crate's type system so that we can safely mark the emulator instances as `Send`, allowing them to be passed between threads. 

2. I added a handful of generic convenience functions such as `write_stack_pointer()` and `read_program_counter()`, that let certain operations be performed in a regular fashion on various CPU architectures. 

The aim in each case has been to make the library more useful to me in my own project (a rewrite of ROPER, at https://github.com/oblivia-simplex/berbalang), where I'm building a system for evolving ROP payloads on various CPU architectures -- something that puts a real premium on both concurrency and genericity. So feel free to cherry-pick these changes as you see fit.

### CODE REVIEW

**Comment by XVilka** [comment]
Looks like this repo is currently unmaintained, see the discussion https://github.com/ekse/unicorn-rs/issues/69

Maybe @lzutao could help to review and merge?

File: Cargo.toml [hunk 1]
```diff
@@ -20,3 +20,8 @@ description = "Rust bindings for the unicorn emulator"
 bitflags = "1.0"
 libc = "0.2"
 libunicorn-sys = { path = "libunicorn-sys", version = "0.9.0" }
+log = "0.4"
+pretty_env_logger = "0.4"
```
Comment (ekse) [inline Cargo.toml:24]: I think pretty_env_logger is not being used.

File: libunicorn-sys/Cargo.toml [hunk 2]
```diff
@@ -26,6 +26,8 @@ pkg-config = "0.3"
 [dependencies]
 bitflags = "1.0"
 libc = { version = "0.2", default-features = false }
+log = "0.4"
+pretty_env_logger = "0.4"
```
Comment (ekse) [inline libunicorn-sys/Cargo.toml:30]: I think pretty_env_logger is not being used.

File: libunicorn-sys/src/unicorn_const.rs [hunk 3]
```diff
@@ -108,6 +108,24 @@ pub struct MemRegion {
...
+        self.perms.intersects(Protection::WRITE)
+    }
+
+    pub fn readable(&self) -> bool {
+        self.perms.intersects(Protection::READ)
+    }
+
+    pub fn executable(&self) -> bool {
+        self.perms.intersects(Protection::EXEC)
+    }
```
Comment (ekse) [inline libunicorn-sys/src/unicorn_const.rs:128]: Please add doc comments for these methods.

File: src/lib.rs [hunk 4]
```diff
@@ -313,47 +341,55 @@ implement_emulator!(
...
 
 implement_emulator!(
     doc = "An X86 emulator instance.",
     doc = "Create an X86 emulator instance for the specified hardware mode.",
     CpuX86,
     Arch::X86,
-    RegisterX86
+    RegisterX86,
+    // special registers
+    // TODO: handle modes
```
Comment (ekse) [inline src/lib.rs]: Is this TODO for something you wanted to add in this pull request? If not I suggest putting a bit more details in the message.

**Comment by ekse** [comment]
The Rust bindings will be packaged directly with unicorn from now on, I will archive this project. I invite you to submit the pull request there.

https://github.com/unicorn-engine/unicorn/tree/next/bindings/rust

**Pull request closed without merging**